### PR TITLE
Adcm-2047 add tests with incorrect hosts in groups

### PR DESCRIPTION
**Pull request by annademina**

### CODE REVIEW

File: tests/functional/test_config_groups.py [hunk 1]
```diff
@@ -28,64 +28,86 @@
...
-
-    @allure.step("Check that the only second host is present in candidates on second provider group")
-    def assert_host_candidate_equal_expected(self, group: HostList, expected_hosts_amount: int):
-        with allure.step(f"Check that {expected_hosts_amount} hosts are available in group"):
-            assert len(group) == expected_hosts_amount, f"{expected_hosts_amount} hosts should be available in group"
-        with allure.step(f"Check that host '{self.SECOND_HOST}' is available in group"):
-            assert group[0].fqdn == self.SECOND_HOST, f"Should be available host '{self.SECOND_HOST}'"
+@allure.step('Check error')
+def assert_that_host_add_is_unavailable(service_group: GroupConfig, host: Host):
+    with allure.step(f"Check that error is '{GROUP_CONFIG_HOST_ERROR.code}'"):
```
Comment (Sealwing) [inline tests/functional/test_config_groups.py:84]: Why code, not error?
Just a suggestion: maybe reverse quotes?

File: tests/functional/test_config_groups.py [hunk 2]
```diff
@@ -28,64 +28,86 @@
...
+    with allure.step(f"Check error message is '{HOST_EXISTS_MESSAGE}'"):
+        assert HOST_EXISTS_MESSAGE in e.value.error['desc'], f"Should be error message '{HOST_EXISTS_MESSAGE}'"
+
 
+@allure.step("Check that the only second host is present in candidates on second provider group")
+def assert_host_candidate_equal_expected(group: HostList, expected_hosts_name: [str]):
+    with allure.step(f"Check that {len(expected_hosts_name)} hosts are available in group"):
+        assert len(group) == len(expected_hosts_name), f"{len(expected_hosts_name)} hosts should be available in group"
+    with allure.step(f"Check that host '{SECOND_HOST}' is available in group"):
+        assert [g.fqdn for g in group] == expected_hosts_name, f"Should be available hosts '{expected_hosts_name}'"
```
Comment (Sealwing) [inline tests/functional/test_config_groups.py:107]: bind `len(expected_hosts_name)` earlier and use it here and later

File: tests/functional/test_config_groups.py [hunk 3]
```diff
@@ -28,64 +28,86 @@
...
+    with allure.step(f"Check that error is '{GROUP_CONFIG_HOST_EXISTS.code}'"):
+        with pytest.raises(ErrorMessage) as e:
+            service_group.host_add(host)
+        GROUP_CONFIG_HOST_EXISTS.equal(e)
+    with allure.step(f"Check error message is '{HOST_EXISTS_MESSAGE}'"):
+        assert HOST_EXISTS_MESSAGE in e.value.error['desc'], f"Should be error message '{HOST_EXISTS_MESSAGE}'"
+
 
+@allure.step("Check that the only second host is present in candidates on second provider group")
+def assert_host_candidate_equal_expected(group: HostList, expected_hosts_name: [str]):
```
Comment (Sealwing) [inline tests/functional/test_config_groups.py:103]: If you use "expected_hosts_name", then it's confusing that it expects list, use "names" instead

File: tests/functional/test_config_groups.py [hunk 4]
```diff
@@ -28,64 +28,86 @@
...
+def assert_that_host_exists(service_group: GroupConfig, host: Host):
+    with allure.step(f"Check that error is '{GROUP_CONFIG_HOST_EXISTS.code}'"):
+        with pytest.raises(ErrorMessage) as e:
+            service_group.host_add(host)
+        GROUP_CONFIG_HOST_EXISTS.equal(e)
+    with allure.step(f"Check error message is '{HOST_EXISTS_MESSAGE}'"):
+        assert HOST_EXISTS_MESSAGE in e.value.error['desc'], f"Should be error message '{HOST_EXISTS_MESSAGE}'"
+
 
+@allure.step("Check that the only second host is present in candidates on second provider group")
```
Comment (skhomuti) [inline tests/functional/test_config_groups.py:102]: this step is common, not only provider

File: tests/functional/test_config_groups.py [hunk 5]
```diff
@@ -110,51 +132,128 @@ def test_that_groups_not_allowed_to_intersect_in_cluster(self, sdk_client_fs, cl
...
+        test_host_1, test_host_2 = create_two_hosts
+        cluster.host_add(test_host_1)
+        cluster.hostcomponent_set(
+            (test_host_1, service.component(name=FIRST_COMPONENT_NAME)),
+            (test_host_1, service.component(name=SECOND_COMPONENT_NAME)),
+        )
+        return service, test_host_1, test_host_2
+
+    def test_add_incorrect_host_to_provider_group(self, sdk_client_fs, provider_bundle, provider):
+        """Test exception rise when we try to add incorrect host to provider group"""
```
Comment (skhomuti) [inline tests/functional/test_config_groups.py:198]: Add to test description what is mean by "incorrect host"

File: tests/functional/test_config_groups.py [hunk 6]
```diff
@@ -110,51 +132,128 @@ def test_that_groups_not_allowed_to_intersect_in_cluster(self, sdk_client_fs, cl
...
+            assert_that_host_exists(provider_group, correct_host)
+            assert_host_candidate_equal_expected(provider_group.host_candidate(), [])
+
+    def test_add_incorrect_host_to_service_group(self, cluster_with_components_on_one_host):
+        """Test exception rise when we try to add incorrect host to service group"""
+
+        service, test_host_1, test_host_2 = cluster_with_components_on_one_host
+        with allure.step("Create group for service"):
+            service_group = service.group_config_create(name=FIRST_GROUP)
+        with allure.step("Create the second group for service and check that not allowed to add the first host to it"):
```
Comment (skhomuti) [inline tests/functional/test_config_groups.py:220]: this is step from previous tests, please check all and fix

**Review by skhomuti — CHANGES_REQUESTED** [review]

**Comment by annademina** [comment]
test_config_groups tests fail until https://github.com/arenadata/adcm_client/pull/87 is merged

File: tests/functional/test_config_groups.py (hunk 1 above)
Comment (annademina) [reply tests/functional/test_config_groups.py:84]: I guess code will describe the problem, ex GROUP_CONFIG_HOST_ERROR

File: tests/functional/test_config_groups.py (hunk 2 above)
Comment (annademina) [reply tests/functional/test_config_groups.py:107]: I dont need len there, this line checks names of hosts

Comment (Sealwing) [reply tests/functional/test_config_groups.py:107]: You use `len(expected_hosts_name)` 3 times here, I've just selected more lines  that needed

File: tests/functional/test_config_groups.py (hunk 1 above)
Comment (Sealwing) [reply tests/functional/test_config_groups.py:84]: Ah, okay, I've mistaken code for title

File: tests/functional/test_config_groups.py (hunk 6 above)
Comment (annademina) [reply tests/functional/test_config_groups.py]: fixed. thnx!

File: tests/functional/test_config_groups.py (hunk 5 above)
Comment (annademina) [reply tests/functional/test_config_groups.py:201]: added to docstring in class

File: tests/functional/test_config_groups.py (hunk 4 above)
Comment (annademina) [reply tests/functional/test_config_groups.py]: fixed

File: tests/functional/test_config_groups.py (hunk 3 above)
Comment (annademina) [reply tests/functional/test_config_groups.py]: changed

File: tests/functional/test_config_groups.py (hunk 2 above)
Comment (annademina) [reply tests/functional/test_config_groups.py]: added variable

File: tests/functional/test_config_groups.py [hunk 7]
```diff
@@ -217,7 +220,7 @@ def test_add_incorrect_host_to_service_group(self, cluster_with_components_on_on
         service, test_host_1, test_host_2 = cluster_with_components_on_one_host
         with allure.step("Create group for service"):
             service_group = service.group_config_create(name=FIRST_GROUP)
-        with allure.step("Create the second group for service and check that not allowed to add the first host to it"):
+        with allure.step("Try to add the second host not from service nd check group hosts list"):
```
Comment (skhomuti) [inline tests/functional/test_config_groups.py:223]: `and` typo

Comment (annademina) [reply tests/functional/test_config_groups.py]: fixed

**Review by Sealwing — APPROVED** [review]

**Review by skhomuti — APPROVED** [review]

**Pull request merged**